### ISSUE TITLE
[sklearn#33960] PERF: Avoid unnecessary n_samples x n_samples diagonal sparse matrix in `_rescale_data`

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#369]
**Mirrored from [scikit-learn#33960](https://github.com/scikit-learn/scikit-learn/issues/33960)**

Original title: PERF: Avoid unnecessary n_samples x n_samples diagonal sparse matrix in `_rescale_data`
Original labels: New Feature, Needs Triage
Created: 2026-05-09T18:13:53Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the workflow you want to enable

When fitting any linear model (e.g. `LinearRegression`, `Ridge`, `ElasticNet`) on sparse data
with `sample_weight`, the internal function `_rescale_data` is called to bake sample weights
into the data before solving. Currently it constructs an intermediate `(n_samples x n_samples)`
diagonal sparse matrix just to perform a simple row scaling, then runs a full sparse matrix
multiply through `safe_sparse_dot`.

For large sparse datasets this creates unnecessary memory allocation and computation on every
`fit()` call. The issue affects all linear models that support `sample_weight` with sparse input.

### Describe your proposed solution

Replace the `dia_array` + `safe_sparse_dot` approach with a direct elementwise row-scaling via
`csr_array.multiply()`, which is built for this exact operation and needs no intermediate matrix.

**Current code** in `sklearn/linear_model/_base.py:252`:

```python
if sp.issparse(X) or sp.issparse(y):
    sw_matrix = sparse.dia_array(
        (sample_weight_sqrt, 0), shape=(n_samples, n_samples)
    )

if sp.issparse(X):
    X = safe_sparse_dot(sw_matrix, X)

if sp.issparse(y):
    y = safe_sparse_dot(sw_matrix, y)
```

**Proposed code:**

```python
if sp.issparse(X):
    X = X.multiply(sample_weight_sqrt[:, None])

if sp.issparse(y):
    sw = sample_weight_sqrt[:, None] if y.ndim > 1 else sample_weight_sqrt
    y = y.multiply(sw)
```

Both produce identical results — verified against the existing `test_rescale_data` test and
directly against the old implementation across all input combinations.

**Why it is better:**

- Eliminates the `(n_samples x n_samples)` diagonal sparse matrix allocation entirely
- `csr_array.multiply(vector)` is O(nnz) with no intermediate object;
  `dia_array @ csr_array` constructs the diagonal object first, then runs a matrix-multiply
  code path
- For 100k samples: old code allocates a 100k-element `data` array and shape tuple for the
  `dia_array` before touching `X`; new code works directly on the `nnz` non-zero entries

## Describe alternatives you've considered

- `scipy.sparse.diags(sample_weight_sqrt) @ X` — same problem, still builds an intermediate
  diagonal sparse matrix
- Keeping `safe_sparse_dot` — unnecessary here since `.multiply()` handles elementwise
  scaling natively and always returns a sparse result

## Additional context

- **Affected function:** `sklearn.linear_model._base._rescale_data` (line 220)
- **Affected models:** all linear models that accept `sample_weight`

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [35](https://github.com/godofecht/flow-scikit/issues/35) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [36](https://github.com/godofecht/flow-scikit/issues/36) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [37](https://github.com/godofecht/flow-scikit/issues/37) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
